### ISSUE TITLE
OOM cause the crash when i decode large PDF vector image.

### ISSUE THREAD

**Issue by yindongwang** [SDWebImage/SDWebImage#3369]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           |  ios 
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | 5.13.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | 13.4
 Repro rate              | all the time (100%)
 Repro with our demo prj | all the time (100%)
 Demo project link       | git@github.com:yindongwang/IssueDemo.git

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

In chinese:
我尝试加载一个.ai的矢量图。当SDImageIOAnimatedCoder在创建image的时候，图片的宽高竟然都超过了10000，最终导致使用imageView加载image时发生OOM。
在调试过程中发现，SDImageIOAnimatedCoder.m的231行出现的kSDCGImageSourceRasterizationDPI（kCGImageSourceRasterizationDPI）并未在ImageIO中定义，不知道作用是什么。
我尝试使用kCGImageSourceThumbnailMaxPixelSize以及通过CGImageSourceCreateThumbnailAtIndex创建图片，可以暂时解决我的问题。

**Comment by dreampiggy** [comment]
This is used to render a vector image into bitmap image. The value is the bitmap pixel size.

**Comment by dreampiggy** [comment]
It seems you are loaing a huge vector image, its own bounding box is really large.

For this case, use `imageThumbnailPixelSize` context option can limit the size.

Or, another case, we don't transform the vector image into bitmap image, and just return nil for this ?

**Comment by yindongwang** [comment]
> It seems you are loaing a huge vector image, its own bounding box (ImageIO tell use from `kCGImagePropertyPixelWidth` and `kCGImagePropertyPixelHeight`) is really large.
> 
> For this case, use `imageThumbnailPixelSize` context option can limit the size. (Or use the `SDWebImageScaleDownLargeImages` option, by default limit pixel to `3968x3968`)
> 
> Or, another case, we don't transform the vector image into bitmap image, and just return nil for this ? Because ImageIO can only provide acutually bitmap format for PDF (not vector image !!)
> 
> Use https://github.com/SDWebImage/SDWebImagePDFCoder is better, pure vector rendering.

Thanks. SDWebImagePDFCoder is really better. 

But I still want to know what `kSDCGImageSourceRasterizationDPI` means ?  `CGImageSourceCreateImageAtIndex` has this option ? Or it is a private options ?

**Comment by yindongwang** [comment]
> 

It works. But the image is very huge (use screen size on iPhone12 pro simulator, and image size is 13926 x 19694). 
Whether it should be use `CGImageSourceCreateThumbnailAtIndex` to create thumbnail image, and  use `imageThumbnailPixelSize ` option ?

**Comment by dreampiggy** [comment]
From the code above, if you don't provide any `imageThumbnailPixelSize`, why this cause `use screen size on iPhone12 pro simulator, and image size is 13926 x 19694)` ? It will enter the `maxPixelHeight == 0 || maxPixelWidth == 0` and use the screen size instead

Or ? Maybe that `isSymbolImage` check failed ? Could you provide the sample image url (or private send email to me)

**Comment by yindongwang** [comment]
Actually, it use the screen size instead. 
```
CGFloat maxPixelSize = MAX(thumbnailSize.width, thumbnailSize.height);
NSUInteger DPIPerPixel = 2;
NSUInteger rasterizationDPI = maxPixelSize * DPIPerPixel;
decodingOptions[kSDCGImageSourceRasterizationDPI] = @(rasterizationDPI);
```
value of `rasterizationDPI` is 1688, after create imageRef, `CGImageGetWidth(imageRef)` return 13926.


I already send mail to you, thanks.

**Comment by dreampiggy** [comment]
Seems bug. The RasterizationDPI value is not what I think for Bitmap image. It's PDF concept.

Use a fixed value is enough. Or I can query the PDF page size and calculate to translate into `pixel size`

See: https://communities.bentley.com/products/microstation/w/microstation__wiki/8867/pdf-file-handling

**Comment by dreampiggy** [comment]
The bug should be solved in #3370.

See test case:

1. When no `imageThumbnailPixelSize` provided, use PDF's own media box rect to generate pixel image
2. When `imageThumbnailPixelSize` provided, calculate the DPI and let the generated pixel image size equal maxPixelSize, always keep aspect ratio.
